### PR TITLE
Add table function `mergeTreeIndex`

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added table function `mergeTreeIndex`. It represents the contents of index and marks files of `MergeTree` tables. It can be used for introspection. Syntax: `mergeTreeIndex(database, table, [with_marks = true])` where `database.table` is an existing table with `MergeTree` engine. 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 57d61a7bc1980c89bea4ae69176e7b84ff0deeb2 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58140/57d61a7bc1980c89bea4ae69176e7b84ff0deeb2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Storages/StorageMergeTreeIndex.cpp [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+        auto & compressed_data = compressed->getData();
+        auto & uncompressed_data = uncompressed->getData();
+
+        for (size_t i = 0; i < num_rows; ++i)
+        {
+            auto mark = marks_loader->getMark(i, col_idx);
+
+            compressed_data[i] = mark.offset_in_compressed_file;
+            uncompressed_data[i] = mark.offset_in_decompressed_block;
```
Comment (alexey-milovidov) [inline src/Storages/StorageMergeTreeIndex.cpp:188]: And row count for adaptive granularity?

Comment (CurtizJ) [reply src/Storages/StorageMergeTreeIndex.cpp:188]: It's already added (column `rows_in_granule`)

**Comment by tavplubix** [comment]
> Test failures are unrelated.

`Stress test (asan) — Found signal in gdb.log`  and `Upgrade check (msan) — Possible deadlock` are still quite interesting, are they already known? 

However, there's another critical issue introduced by this PR, and I failed to fix it quickly, so I have to revert, sorry :(

**Pull request merged**